app/vmselect/netstorage/netstorage.go

```diff
@@ -1716,24 +1719,30 @@ func (sn *storageNode) execOnConn(qt *querytracer.Tracer, funcName string, f fun
 		_ = bc.Close()
 		return fmt.Errorf("cannot send timeout=%d for funcName=%q to the server: %w", timeout, funcName, err)
 	}
+	
 	// Execute the rpc function.
-	if err := f(bc); err != nil {
-		remoteAddr := bc.RemoteAddr()
-		var er *errRemote
-		if errors.As(err, &er) {
-			// Remote error. The connection may be re-used. Return it to the pool.
-			_ = readTrace(qt, bc)
-			sn.connPool.Put(bc)
+	// Remote error. The connection may be re-used. Return it to the pool.
+	var er *errRemote
+	if err := f(bc); err != nil && !errors.As(err, &er) {
+		// Deadline has been introduced as an error instead of Timeout
+		if errors.As(err, os.ErrDeadlineExceeded) {
```

Comment:
It looks like the `errors.Is()` must be used here instead of `errors.As`. See https://pkg.go.dev/errors#Is .